In |filter(), |map(), and |reduce(), throw a RuntimeError instead of a TypeError

thijzert opened the pull request:
If a template accidentally performs `|filter(...)` on a variable that isn't traversable (e.g. `null`), the calling code will encouter the raw `TypeError` that results from Twig trying to construct an `IteratorIterator` on that variable. This means that context information (e.g., in which exact template the error occurred, and on which line) is discarded, which makes it very difficult to debug your templates.

This commit adds code to throw a `RuntimeError` if Twig encounters a situation where there's no valid way to filter an array.

thijzert committed:
In |filter(), |map(), and |reduce(), throw a RuntimeError instead of a TypeError

guilliamxavier commented:
There's already #3326 that has been merged in `1.x` but not into `2.x`/`3.x`.

Also, what about `twig_array_map` and `twig_array_reduce`?

thijzert commented:
I guess I should use `|reduce()` more - it has the same problem. I've added a fix for it.

`twig_array_map` already fails with a RuntimeError in a way that's consistent with my subjective expectations, so I've added a unit test to fix the behaviour in place.

guilliamxavier commented:
Actually `twig_array_map(null, fn ($x) => $x)` doesn't throw an exception per se, just emits a Warning "Invalid argument supplied for foreach()", that I suppose is then converted to a RuntimeError by some error handler... But indeed not a TypeError either way ;)

Anyway `$array instanceof \Traversable || $array instanceof \Iterator` is redundant since `interface Iterator extends Traversable` (see e.g. `twig_array_column`)

thijzert commented:
> Actually `twig_array_map(null, fn ($x) => $x)` doesn't throw an exception per se, just emits a Warning "Invalid argument supplied for foreach()", that I suppose is then converted to a RuntimeError by some error handler.

Ah, that makes sense.

guilliamxavier commented on src/Extension/CoreExtension.php:
```diff
@@ -1573,6 +1573,10 @@ function twig_array_filter($array, $arrow)
     }
 
+    if (!$array instanceof \Traversable) {
+        throw new RuntimeError(sprintf('An array filter only works with arrays or "Traversable", got "%s" as first argument.', \gettype($array)));
```
```suggestion
        throw new RuntimeError(sprintf('The filter filter only works with arrays or "Traversable", got "%s" as first argument.', \gettype($array)));
```
(or maybe
```suggestion
        throw new RuntimeError(sprintf('The "filter" filter only works with arrays or "Traversable", got "%s" as first argument.', \gettype($array)));
```
to mitigate the repetition)

guilliamxavier commented on src/Extension/CoreExtension.php:
```diff
@@ -1590,6 +1594,10 @@ function twig_array_map($array, $arrow)
 {
     if (!\is_array($array)) {
+        if (!$array instanceof \Traversable) {
+            throw new RuntimeError(sprintf('An array filter only works with arrays or "Traversable", got "%s" as first argument.', \gettype($array)));
```
```suggestion
            throw new RuntimeError(sprintf('The reduce filter only works with arrays or "Traversable", got "%s" as first argument.', \gettype($array)));
```

fabpot commented:
Can you rebase on 1.x and remove the check in twig_array_filter as it is redundant?

thijzert commented:
Some would say "redundant," some would say "preemptive regression test," but sure, no problem.

fabpot commented on src/Extension/CoreExtension.php:
```diff
@@ -1707,6 +1707,10 @@ function twig_array_filter($array, $arrow)
         return array_filter($array, $arrow);
     }
 
+    if (!$array instanceof \Traversable) {
```
This test must be removed, the same test is already done line 1698.

thijzert replied:
Done.

fabpot commented:
@thijzert I made a comment to make myself clearer. The test was fine and can be re-added :)

thijzert commented:
Ah, yes, I see the confusion now.

fabpot commented:
Thank you @thijzert.
